streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -122,52 +127,64 @@ public void register(final StateStore store,
+        if (registeredStores.containsKey(storeName)) {
             throw new IllegalArgumentException(String.format("%sStore %s has already been registered.", logPrefix, storeName));
         }
 
         // check that the underlying change log topic exist or not
         final String topic = storeToChangelogTopic.get(storeName);
         if (topic == null) {
-            stores.put(storeName, store);
-            return;
-        }
+            registeredStores.put(storeName, store);
+        } else {
 
-        final TopicPartition storePartition = new TopicPartition(topic, getPartition(topic));
+            final TopicPartition storePartition = new TopicPartition(topic, getPartition(topic));
 
-        final RecordConverter recordConverter = converterForStore(store);
+            final RecordConverter recordConverter = converterForStore(store);
 
-        if (isStandby) {
-            log.trace("Preparing standby replica of persistent state store {} with changelog topic {}", storeName, topic);
+            if (isStandby) {
+                log.trace("Preparing standby replica of persistent state store {} with changelog topic {}", storeName, topic);
 
-            restoreCallbacks.put(topic, stateRestoreCallback);
-            recordConverters.put(topic, recordConverter);
-        } else {
-            log.trace("Restoring state store {} from changelog topic {} at checkpoint {}", storeName, topic, checkpointableOffsets.get(storePartition));
-
-            final StateRestorer restorer = new StateRestorer(
-                storePartition,
-                new CompositeRestoreListener(stateRestoreCallback),
-                checkpointableOffsets.get(storePartition),
-                offsetLimit(storePartition),
-                store.persistent(),
-                storeName,
-                recordConverter
-            );
-
-            changelogReader.register(restorer);
+                restoreCallbacks.put(topic, stateRestoreCallback);
+                recordConverters.put(topic, recordConverter);
+            } else {
+                log.trace("Restoring state store {} from changelog topic {} at checkpoint {}", storeName, topic, checkpointableOffsets.get(storePartition));
+
+                final StateRestorer restorer = new StateRestorer(
+                    storePartition,
+                    new CompositeRestoreListener(stateRestoreCallback),
+                    checkpointableOffsets.get(storePartition),
+                    offsetLimit(storePartition),
+                    store.persistent(),
+                    storeName,
+                    recordConverter
+                );
+
+                changelogReader.register(restorer);
+            }
+            changelogPartitions.add(storePartition);
+
+            registeredStores.put(storeName, store);
         }
-        changelogPartitions.add(storePartition);
 
-        stores.put(storeName, store);
+        ensureStoresIsTopologicallyOrdered();
```

Comment:
The alternative here is to seek to the right spot in the list and insert the new store there, but it seemed easier to read this way. Please share your thoughts.

Reply:
Hmm I'm still not clear where did we break the topology order here: let me go through my reasoning and lmk where I got it wrong:

1. in `InternalTopologyBuilder` when we construct the `InternalTopology` the following parameter is constructed:

```
new ArrayList<>(stateStoreMap.values()),
```

So `ProcessorTopology#stateStores()` is in order.

2. in `AbstractTask#registerStateStores` we get stores from `ProcessorTopology#stateStores()` which is in order, and hence we are calling `store.init` in order, and hence call `ProcessorStateManager#register` in order as well.

3. The resulted `stores` in `ProcessorStateManager` should be in order then as well.

Reply:
I'll investigate it and let you know. I also wondered about this, but didn't look into it, since there's no guarantee that the `register` call takes place synchronously inside of `init`, so if we have a correctness requirement to flush in topo order, we should just flush in topo order, rather than depend on stores being registered in topo order, since that leaves more room for error.

Reply:
Oh, actually I see that `register` is only allowed during initialization... I'll have to go through the whole repro again to track this down. I'll let you know.

Reply:
Ah. The problem is in `org.apache.kafka.streams.processor.internals.AbstractStateManager#reinitializeStateStoresForPartitions`. We're iterating over a copy of the state stores, but when we make the copy, we don't preserve order:

```
        final Map<String, StateStore> storesCopy = new HashMap<>(stateStores);
        ...
        for (final Map.Entry<String, StateStore> entry : storesCopy.entrySet()) {
                ...
                stateStore.init(processorContext, stateStore);
                ...
        }
```

So we have alternative fix, where we fix the order of _re_-initialization, but it's not super straightforward, because not all the stores in the task get re-initialized, so even with an ordered copy of the map, it'll still wind up scrambling the stores. So, we still have the problem of explicitly maintaining topological order in the state manager.

What do you think, should we still use `ensureStoresIsTopologicallyOrdered`, or instead try to structure the initialization/re-initialization so that everything gets registered in the right order?

Reply:
I see. So it seems this issue will only be exposed with EOS still, right (reinitializeStateStoresForPartitions is only triggered when EOS is enabled and offset out-of-range detected)?

As for the fix: since the `storesCopy = new HashMap<>(stateStores)` is just for fixing concurrent access exception, how about we just use an iterators instead, i.e.:

```
iterator = ..
for (iterator) {
  if (storeToBeReinitialized.contains(..)) {
    ...
    iterator.remove()
    stateStore.init(processorContext, stateStore);
  }
}
```

Reply:
We can do that, but I think it will still re-order the stores:
```
Stores: [A, B]
storesToBeReinitialized = [A]

iterator.next: A
storeToBeReinitialized.contains(A) == true
iterator.remove
// now stores: [B]
stateStore.init(processorContext, A)
// now stores: [B, A]
```

This is what I meant by saying that we still have the problem of maintaining topological order during registration, even if we correct the ordering of `storesCopy`. We could add some mechanism to keep it in the same position in the linked map, but part of me feels like the current code in the PR is better, since it just enforces the desired invariant on the stores after every registration, regardless of what order things happen in.

Reply:
I see your point now, this is exactly the messy code that we were trying to fix.

I've looked at the source code again, and I think we can actually not remove the state at all since the same object will be add to the state stores via `store.init` immediately within the same function call.

So I think we can actually do:

```
if (storeToBeReinitialized.contains(A)) {
A.close;
delete state dir;
A.init();
}
```

In that loop.

Reply:
Yes, that would work.

Reply:
Ok, please see my latest comment below. The fly in the ointment was that we don't allow to register a store twice, and the only way we know whether a store is registered is its presence in this map.